src/App.tsx

```diff
@@ -1,26 +1,314 @@
-/* eslint-disable max-len */
 /* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+/* eslint-disable jsx-a11y/label-has-associated-control */
+import React, { useEffect, useRef, useState } from 'react';
+import classNames from 'classnames';
 import { UserWarning } from './UserWarning';
+import { addTodo, deleteTodo, getTodos, USER_ID } from './api/todos';
+import { Todo } from './types/Todo';
+import { client } from './utils/fetchClient';
+import { Header } from './components/Header';
+import { TodoList } from './components/TodoList';
+import { Footer } from './components/Footer';
 
-const USER_ID = 0;
+enum ErrorMessages {
+  LOAD_TODOS = 'Unable to load todos',
+  ADD_TODO = 'Unable to add a todo',
+  DELETE_TODO = 'Unable to delete a todo',
+  UPDATE_TODO = 'Unable to update a todo',
+  EMPTY_TITLE = 'Title should not be empty',
+}
 
 export const App: React.FC = () => {
+  const [todosList, setTodosList] = useState<Todo[]>([]);
+  const [currentFilter, setCurrentFilter] = useState('All');
+  const [errorMsg, setErrorMsg] = useState('');
+
+  const [newTitle, setNewTitle] = useState('');
+  const newTitleRef = useRef<HTMLInputElement>(null);
+  const [pendingTodo, setPendingTodo] = useState<Todo | null>(null);
+
+  const [deletingTodoId, setDeletingTodoId] = useState<number | null>(null);
+  const [deletingTodosIds, setDeletingTodosIds] = useState<number[]>([]);
+
+  const [updatingIds, setUpdatingIds] = useState<number[]>([]);
+  const [editingTodoId, setEditingTodoId] = useState<number | null>(null);
+  const [editingTitle, setEditingTitle] = useState('');
+
+  const remainingItems = todosList.filter(todo => !todo.completed);
+
+  const isAddingTodo = pendingTodo !== null;
+
+  const updateTodo = (id: number, data: Partial<Todo>) => {
+    return client.patch<Todo>(`/todos/${id}`, data);
+  };
+
+  useEffect(() => {
+    newTitleRef.current?.focus();
+    const fetchTodos = async () => {
+      try {
+        const todos = await getTodos();
+
+        setTodosList(todos);
+      } catch {
+        setErrorMsg(ErrorMessages.LOAD_TODOS);
+      }
+    };
+
+    fetchTodos();
+  }, []);
+
+  useEffect(() => {
+    if (editingTodoId !== null) {
+      const input = document.querySelector<HTMLInputElement>(
+        `[data-cy="TodoTitleField"]`,
+      );
+
+      input?.focus();
+    }
+  }, [editingTodoId]);
+
+  useEffect(() => {
+    if (!errorMsg) {
+      return;
+    }
+
+    const timer = setTimeout(() => setErrorMsg(''), 3000);
+
+    return () => clearTimeout(timer);
+  }, [errorMsg]);
+
+  const filterTodos = (todos: Todo[], filter: string) => {
+    if (filter === 'Active') {
```

Comment:
still using magic string - we should fix it - use enum values instead of magic string